### PR TITLE
BB implementation

### PR DESCRIPTION
**Pull request by Stan1slavssKy**

### CODE REVIEW

File: simulator/hart/basic_block.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    size_t curr_pc = start_pc;
+
+    while (true) {
+        hart_->FetchInstruction(&raw_instr);
+
+        Instruction *instr = bb->AddInstruction();
+        hart_->DecodeInstruction(instr, raw_instr);
+        instr->pc = curr_pc;
+
+        if (instr->IsBranch()) {
```
Comment (Vokerlee) [inline simulator/hart/basic_block.cpp:33]: May be use UNLIKELY()?

File: simulator/hart/basic_block.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    instr_size_t raw_instr = 0;
+
+    size_t curr_pc = start_pc;
+
+    while (true) {
+        hart_->FetchInstruction(&raw_instr);
+
+        Instruction *instr = bb->AddInstruction();
+        hart_->DecodeInstruction(instr, raw_instr);
+        instr->pc = curr_pc;
```
Comment (Vokerlee) [inline simulator/hart/basic_block.cpp:31]: Is it necessary?

File: simulator/hart/hart.cpp [hunk 3]
```diff
@@ -84,35 +88,93 @@ Exception Hart::FetchInstruction(instr_size_t *raw_instr) const
...
+            handlers_.mmu_handler(exception, vaddr_t(instructions[instr_idx].pc));      \
+        }                                                        \
+        pc_ = pc_target_; \
+        DISPATCH();
+
+    // All basic blocks ends with branch instruction
+    #define DEFINE_BRANCH_INSTR(instr)                           \
+    instr:                    \
+        std::cout << "branch instr:" << #instr << " " << static_cast<int>(instructions[instr_idx].id) << std::endl; \
+        DumpRegs(std::cerr);                                    \
```
Comment (Vokerlee) [inline simulator/hart/hart.cpp:155]: To delete

File: simulator/hart/hart.cpp [hunk 4]
```diff
@@ -84,35 +88,93 @@ Exception Hart::FetchInstruction(instr_size_t *raw_instr) const
...
+        DumpRegs(std::cerr); \
+        exception = iexec::instr(this, instructions[instr_idx]); \
+        DumpRegs(std::cerr); \
+        if (UNLIKELY(exception != Exception::NONE)) {            \
+            handlers_.mmu_handler(exception, vaddr_t(instructions[instr_idx].pc));      \
+        }                                                        \
+        pc_ = pc_target_; \
+        DISPATCH();
+
+    // All basic blocks ends with branch instruction
```
Comment (Vokerlee) [inline simulator/hart/hart.cpp:151]: ends -> end

File: simulator/hart/hart.cpp [hunk 5]
```diff
@@ -84,35 +88,93 @@ Exception Hart::FetchInstruction(instr_size_t *raw_instr) const
...
+    };
+
+    #undef DEFINE_INSTR
+    #undef DEFINE_BRANCH_INSTR
+
+    const auto &instructions = bb->GetInstructions();
+    size_t instr_idx = 0;
+    Exception exception = Exception::NONE;
+
+    for (auto it: instructions) {
```
Comment (Vokerlee) [inline simulator/hart/hart.cpp:130]: To delete

**Review by Vokerlee — CHANGES_REQUESTED** [review]

File: simulator/hart/basic_block.h [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+private:
+    // <start_pc, BasicBlock>
+    std::vector<BasicBlock *> cache_;
+};
+
+class BasicBlockManager {
+public:
+    static constexpr size_t DEFAULT_BB_CACHE_SIZE = 1 << 8;
```
Comment (Vokerlee) [inline simulator/hart/basic_block.h:91]: Add to `common/config.h`, please

**Review by Vokerlee — CHANGES_REQUESTED** [review]

File: simulator/hart/basic_block.h [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        start_pc_ = pc;
+    }
+
+    size_t GetStartPC() const
+    {
+        return start_pc_;
+    }
+
+    Instruction *AddInstruction()
```
[the remark this answers is not in the document]
Comment (Stan1slavssKy) [reply simulator/hart/basic_block.h]: done

File: simulator/hart/basic_block.h [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    NO_COPY_SEMANTIC(BasicBlockCache);
+    NO_MOVE_SEMANTIC(BasicBlockCache);
+
+    explicit BasicBlockCache() : cache_(CACHE_SIZE) {}
+
+    ~BasicBlockCache() = default;
+
+    BasicBlock *GetBBFromCache(size_t pc) const
+    {
+        BasicBlock *bb = cache_[pc % CACHE_SIZE];
```
[the remark this answers is not in the document]
Comment (Stan1slavssKy) [reply simulator/hart/basic_block.h]: done

File: simulator/hart/basic_block.h [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        return (bb->GetStartPC() == pc) ? bb : nullptr;
+    }
+
+    void SetBBIntoCache(BasicBlock *bb)
+    {
+        assert(bb != nullptr);
+        size_t start_pc = bb->GetStartPC();
+        // It may happen that cache_ already has an entry in the start_pc % CACHE_SIZE index
+        // In this case entry just pushed out in favor of a new one
+        cache_[start_pc % CACHE_SIZE] = bb;
```
[the remark this answers is not in the document]
Comment (Stan1slavssKy) [reply simulator/hart/basic_block.h]: done

File: simulator/hart/basic_block.cpp (hunk 1 above)
Comment (Stan1slavssKy) [reply simulator/hart/basic_block.cpp]: In this case it is bad

File: simulator/hart/basic_block.cpp (hunk 2 above)
Comment (Stan1slavssKy) [reply simulator/hart/basic_block.cpp]: No, done

File: simulator/hart/hart.cpp (hunk 3 above)
Comment (Stan1slavssKy) [reply simulator/hart/hart.cpp]: done

File: simulator/hart/hart.cpp (hunk 4 above)
Comment (Stan1slavssKy) [reply simulator/hart/hart.cpp]: OK

File: simulator/hart/hart.cpp (hunk 5 above)
Comment (Stan1slavssKy) [reply simulator/hart/hart.cpp]: OK

File: simulator/hart/basic_block.h (hunk 6 above)
Comment (Stan1slavssKy) [reply simulator/hart/basic_block.h]: done

File: simulator/hart/basic_block.h [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+class BasicBlockCache {
+public:
+    NO_COPY_SEMANTIC(BasicBlockCache);
+    NO_MOVE_SEMANTIC(BasicBlockCache);
+
+    explicit BasicBlockCache() : cache_(CACHE_SIZE) {}
+
+    ~BasicBlockCache() = default;
+
+    BasicBlock *GetBBFromCache(size_t pc) const
```
[the remark this answers is not in the document]
Comment (Stan1slavssKy) [reply simulator/hart/basic_block.h]: I think its normal that in Interpret(), basic block manager will return non-const BasibBlock from GetNextBB()

**Comment by Stan1slavssKy** [comment]
@Vokerlee PTAL

**Comment by Stan1slavssKy** [comment]
@Hollbrok PTAL

**Review by Vokerlee — APPROVED** [review]

**Review by Hollbrok — APPROVED** [review]
great work

**Pull request merged**